### PR TITLE
Address DASC review feedback

### PR DESCRIPTION
**Pull request by kaix-nv**
## Summary

Consolidates the complete reviewed fix stack for #2375 into one DCO-safe commit:

- harden package exports, measurement semantics, wrapper handling, and actionable calibration errors
- validate exact installed GDN identities plus ModelOpt dynamic subclasses; reject lookalikes, ordinary subclasses, incomplete layers, and partial layer sets
- make stale checkpoints saveable and restorable while keeping deployment export strict
- make DASC recalibration replace and deduplicate existing mode state without stale-metadata refresh
- record the declared decay-parameter checkpoint storage dtype and use derived FP16/BF16/FP32 rounding bounds
- preserve BF16/FP16 storage and wider/cross-dtype reload compatibility without globally widening FP32 tolerance
- add installed Transformers path coverage, optional Megatron gating, lifecycle, tamper, lossy-cast, and mixed-layer regressions
- document the explicit storage-dtype contract

This consolidated PR supersedes the mechanically stacked review-fix PRs #2377, #2378, #2379, #2380, #2382, #2383, #2384, and #2385. Its tree is byte-identical to the independently reviewed leaf commit from #2386.

## Validation

- focused DASC suite: 23 passed, 1 absent optional Megatron skip
- DASC plus weight sparsity plus attention sparsity compatibility suite: 134 passed, 1 optional skip
- DASC package coverage: 408/408 statements, 100%
- full pre-commit on all touched files: passed
- real Transformers Qwen3NextGatedDeltaNet BF16 storage to FP32 reload smoke: passed
- commit author and Signed-off-by identity both use kaix-nv <kaix@nvidia.com>

## Summary by CodeRabbit

* **New Features**
  * Added support for configuring decay-parameter storage precision with FP16, BF16, or FP32.
  * Added safer recalibration that replaces existing DASC state.
  * Expanded compatibility with supported GDN adapter classes and model wrappers.
  * Added improved validation for sparsity policies, measurements, model structure, and decay parameters.
  * Added support for perplexity-retention values above 1.

* **Documentation**
  * Clarified evaluation responsibilities, recalibration behavior, stale-policy handling, supported adapters, and dtype requirements.

### CODE REVIEW

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 1]
```diff
@@ -191,8 +347,8 @@ def build_dasc_policy(
     validated_measurements.sort(key=lambda measurement: measurement.wmax)
 
     modules = _get_gdn_modules(model)
-    horizons = analyze_gdn_decay(
-        model, epsilon=config.epsilon, static_gate_input=config.static_gate_input
+    horizons = _analyze_gdn_modules(
+        modules, epsilon=config.epsilon, static_gate_input=config.static_gate_input
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py]: Valid. Fixed in #2388: calibration now derives horizons, measurement geometry, and retained masks from A_log/dt_bias canonicalized to the declared storage dtype. The regression pins an FP16 cast that crosses the strict Wmax=7 boundary.

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 2]
```diff
@@ -245,14 +404,48 @@ def build_dasc_policy(
...
+        static_gate_input=policy.static_gate_input,
+    )
+    for name, values in current_horizons.items():
+        layer = policy.layers[name]
+        retained = [head for head, horizon in enumerate(values) if horizon > policy.selected_wmax]
+        if retained != layer.retained_heads:
+            raise ApplyModeError(
+                f"DASC policy head mask does not match current decay parameters in layer {name!r}"
+            )
+        lower, upper = _storage_cast_horizon_bounds(
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py]: Valid. Fixed in #2388 with the combined contract: calibration uses storage-canonical values, while export accepts a mask only when the inverse cast interval can explain the boundary side. It still rejects a declared retained/omitted choice when the full admissible interval lies on the opposite side of strict horizon > Wmax.

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 3]
```diff
@@ -16,19 +16,65 @@
 """Decay analysis and policy selection for GDN state sparsity."""
 
 import hashlib
+import importlib
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py:19]: Valid. Fixed in #2388 by importing importlib.util explicitly; the graceful missing-framework path no longer depends on a transitive import side effect.

File: docs/source/guides/6_sparsity.rst [hunk 4]
```diff
@@ -170,6 +174,16 @@ ordinary dense recurrent state before continuation.
...
+The initial GDN adapter accepts the ``GatedDeltaNet`` and ``Qwen3NextGatedDeltaNet`` base classes,
+including ModelOpt-generated dynamic subclasses, and fails closed for unrelated implementations
+even when they expose similarly named decay tensors.
+Re-running :func:`~modelopt.torch.sparsity.state_sparsity.calibrate` replaces the existing DASC
+mode-state entry and supersedes its stale policy without growing the checkpoint history. A stale
+policy remains serializable so it cannot block saving or composing other ModelOpt modes, but
+:func:`~modelopt.torch.sparsity.state_sparsity.export_policy` rejects it until recalibration.
+Set ``decay_parameter_storage_dtype`` to the checkpoint dtype for ``A_log`` and ``dt_bias`` before
+calibration. Policy validation allows only the rounding introduced by that declared storage dtype
+and the live tensor dtype; the default ``float32`` keeps unconfigured policies strict.
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply docs/source/guides/6_sparsity.rst]: Valid. Fixed in #2388. The guide now states that effective validation tolerance includes both declared storage and live tensor dtypes, and that default float32 adds no storage slack of its own.

File: modelopt/torch/sparsity/state_sparsity/conversion.py [hunk 5]
```diff
@@ -82,25 +91,65 @@ def restore_dasc_model(model: nn.Module, config: DASCConfig, metadata: MetadataD
...
 
-    validate_dasc_model_structure(model, policy)
+    try:
+        validate_dasc_model_structure(model, policy)
+    except ApplyModeError as error:
+        warnings.warn(
+            f"{error}. The restored DASC policy is stale; re-run calibrate() before deployment",
+            stacklevel=2,
+        )
     _attach_policy(model, policy)
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/conversion.py:106]: Valid. Fixed in #2388 by distinguishing recoverable GDN geometry drift from unsupported architecture. Restore now warns only for the former and fails closed when the target has no supported GDN modules; a regression covers restoring onto nn.Linear.
